### PR TITLE
tests: Add fuzzing harnesses for functions in script/

### PR DESCRIPTION
**Pull request by practicalswift**
Add fuzzing harnesses for functions in `script/`:
* Add fuzzing helper functions `ConsumeDataStream` and `ConsumeUInt160`
* Fill fuzzing coverage gaps for functions in `script/script.h`, `script/script_error.h` and `script/standard.h`
* Add fuzzing harness for functions in `script/bitcoinconsensus.h`
* Add fuzzing harness for functions in `script/descriptor.h`
* Add fuzzing harness for functions in `script/interpreter.h`
* Add fuzzing harness for functions in `script/sigcache.h`
* Add fuzzing harness for functions in `script/sign.h`

See [`doc/fuzzing.md`](https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md) for information on how to fuzz Bitcoin Core. Don't forget to contribute any coverage increasing inputs you find to the [Bitcoin Core fuzzing corpus repo](https://github.com/bitcoin-core/qa-assets).

Happy fuzzing :)

### CODE REVIEW

File: src/test/fuzz/script.cpp [hunk 1]
```diff
@@ -119,4 +125,41 @@ void test_one_input(const std::vector<uint8_t>& buffer)
             wit.SetNull();
         }
     }
+
+    const char* op_name = GetOpName(ConsumeOpcodeType(fuzzed_data_provider));
+    assert(op_name != nullptr);
+
+    const char* error_string = ScriptErrorString(static_cast<ScriptError>(fuzzed_data_provider.ConsumeIntegralInRange<int>(0, SCRIPT_ERR_ERROR_COUNT)));
+    assert(error_string != nullptr);
```
Comment (MarcoFalke) [inline src/test/fuzz/script.cpp:133]: unrelated nit: At compile time this can never be null. I'd prefer if the return value for strings was always `std::string`. Every caller of `ScriptErrorString` converts to `std::string` anyway.

File: src/test/fuzz/script.cpp [hunk 2]
```diff
@@ -119,4 +125,41 @@ void test_one_input(const std::vector<uint8_t>& buffer)
...
+        witness_unknown_2.length = witness_unknown_program_2.size();
+        std::copy(witness_unknown_program_2.begin(), witness_unknown_program_2.end(), witness_unknown_2.program);
+
+        (void)(witness_unknown_1 == witness_unknown_2);
+        (void)(witness_unknown_1 < witness_unknown_2);
+
+        const CNoDestination no_destination_1;
+        const CNoDestination no_destination_2;
+        (void)(no_destination_1 == no_destination_2);
+        (void)(no_destination_1 < no_destination_2);
```
Comment (MarcoFalke) [inline src/test/fuzz/script.cpp:163]: this looks like a unit test that does not depend on any fuzz input

File: src/test/fuzz/script_sign.cpp [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+            }
+            std::map<COutPoint, Coin> coins;
+            while (fuzzed_data_provider.ConsumeBool()) {
+                const Optional<COutPoint> outpoint = ConsumeDeserializable<COutPoint>(fuzzed_data_provider);
+                if (!outpoint) {
+                    break;
+                }
+                const Optional<Coin> coin = ConsumeDeserializable<Coin>(fuzzed_data_provider);
+                if (!coin) {
+                    break;
```
Comment (MarcoFalke) [inline src/test/fuzz/script_sign.cpp:134]: unrelated, but I was wondering if `ConsumeDeserializable<T>` could always return type `T` (if the stream was empty, it would return `T{}`). Not returning optional would bring this helper in line with all other helpers. For example `ConsumeRandomLengthString` always return `std::string`.

**Review by maflcko — APPROVED** [review]
ACK

File: src/test/fuzz/script.cpp (hunk 2 above)
Comment (MarcoFalke) [reply src/test/fuzz/script.cpp:163]: My suggestion would be to construct a random `CTxDestination`.

File: src/test/fuzz/script.cpp (hunk 1 above)
Comment (practicalswift) [reply src/test/fuzz/script.cpp:133]: I agree, we should always return `std::string`. Will do that in a follow-up PR since that will touch validation code and thus need more scrutiny than adding tests.

File: src/test/fuzz/script.cpp (hunk 2 above)
Comment (practicalswift) [reply src/test/fuzz/script.cpp]: Done! :)

File: src/test/fuzz/script_sign.cpp (hunk 3 above)
Comment (practicalswift) [reply src/test/fuzz/script_sign.cpp:134]: You mean something along the lines of:

```c++
template <typename T>
NODISCARD inline T ConsumeDeserializable(FuzzedDataProvider& fuzzed_data_provider, const size_t max_length = 4096) noexcept
{
    const std::vector<uint8_t> buffer = ConsumeRandomLengthByteVector(fuzzed_data_provider, max_length);
    CDataStream ds{buffer, SER_NETWORK, INIT_PROTO_VERSION};
    T obj;
    try {
        ds >> obj;
    } catch (const std::ios_base::failure&) {
    }
    return obj;
}
```

I've thought about that and while it would be better from a developer ergonomics perspective I think it might come with some negative impact on overall fuzzing. One risk I see is that the fuzzer might try to work also on "meaningless" inputs (inputs which cause `std::ios_base::failure`) which would now become "somewhat meaningful" due to default construction.

Take the extreme example of a huge input containing a few million bytes worth of 0x41 scream (`AAAAAAAAAAA…`) being deserialized to an object of type `CFoo`. The current fuzzer would quickly reach `std::ios_base::failure`, return a `nullopt` and then give up without further processing (fail early thanks to `if (!opt_foo) { return; }` or similar). If we were to return a `CFoo{}` instead then the fuzzer would proceed with processing (fail late) which may slow down the fuzzing.

Does it make sense? :)

tl;dr -- would be worth doing if no negative fuzzing speed impact can be measured :)

Comment (MarcoFalke) [reply src/test/fuzz/script_sign.cpp:134]: I think when it comes to trading off giving the fuzz test more freedom to explore more code branches vs micro-optimizing for performance, we should always pick the option that gives the fuzz engine more freedom. Fuzzing is great to complement traditional tests (e.g. our unit and functional tests). Those traditional tests offer a reasonable line coverage, but fuzz testing can offer great branch coverage. Giving the fuzz engine to explore as much branches as possible is vital. With modern coverage guided fuzz engines, the strength of fuzzing is that even with a large search space, the fuzz engine will always find useful inputs that increase coverage. Additional search space may (or may not!) come with a performance penalty, but I think we shouldn't use performance as a way to measure usefulness of a fuzz tests. The fastest fuzz test is the one that tests nothing. However, it is also the most useless one. Anything that can be solved by letting the fuzz engine run for another hour or by simply adding one more core should not be a reason to cripple the test into one that yields less coverage.

In fact, this has shown to be effective in the past. Take for example the process_message harness vs the process_message*s* harness. The latter is allowed to send an arbitrary number of messages and start an arbitrary number of peers. The latter (more flexible and broader) test that gives the engine more freedom is also the *only* test that has found bugs in pull requests that none of the other fuzz test or unit test or even functional tests have found. Specifically:

* It found a nullptr dereference: https://github.com/bitcoin/bitcoin/pull/18808#discussion_r417307258
* It found an unidentified problem: https://github.com/bitcoin/bitcoin/pull/15197#pullrequestreview-408731734
* Some more that I am missing...?

Comment (practicalswift) [reply src/test/fuzz/script_sign.cpp:134]: I agree regarding the coverage vs performance trade-off: I'd choose "explore more code branches" over performance any day in week :)

Do you see a scenario where returning a default constructed object in case of serialisation failure is likely to give more freedom to explore more code branches compared to not doing so?

If so we'd have _two_ reasons to do this: developer ergonomics (no more `std::optional`) _and_ increased coverage :)

My thinking is more along the lines of "how do we guide the coverage-guided fuzzer in a meaningful way" rather than "micro-optimizing for performance". I guess the question boils down to if and to what extent the guidance provided by failing early is meaningful for the fuzzer :)

Code if anyone wants to experiment: https://github.com/bitcoin/bitcoin/compare/master...practicalswift:consume

Comment (MarcoFalke) [reply src/test/fuzz/script_sign.cpp:134]: > Do you see a scenario where returning a default constructed object in case of serialisation failure is likely to give more freedom to explore more code branches compared to not doing so?

Unclear. This question boils down to whether default initialization leaves members uninitialized or initializes them with values that can not be the result of deserialization.

Comment (practicalswift) [reply src/test/fuzz/script_sign.cpp:134]: Agreed. Sounds like we agree on the need for experimentation and measuring before proceeding with this :)

**Comment by MarcoFalke** [comment]
Needs rebase

**Comment by MarcoFalke** [comment]
ACK f898ef65c947776750e49d050633f830546bbdc6 🔉

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK f898ef65c947776750e49d050633f830546bbdc6 🔉
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgtcwwAk2++KR5/UnaZ6/Gls4rZq7sjVolrOPUw/ADuxRyFu/qM7nmAOuzhX645
/tuQPOszycpK7MuS8INKbw2gV4TTP35W7N9z8GhtbyaSlRXzBGF5vLGk0P1diJrp
GwSiwUle6VU4sRLX44DctSgP9NvNMBkl+zqFEJpjspXkGZuYvtDJrP4Ipm+oSNOw
d5K0zg6VKwz/+23qt9OJErovRltC7QUxIecI+Y0KmJzfkpScf+jGoMTFgo9q0zC4
JDLJ0ZAU8BeiN2J/IK0yyxG1LhZPOK1LikFWqebeD4JEOSA+FnWopXHCUAj+sL7H
hRZmjH1CJ8my2d2nOdqCmOn1R3K+kEuVB0L2cMpPsISRcPjgI8GlF4RcY23PpTBJ
FtjGQdg30fzpNeLW0aPvti4RV7e8eDsi87BOWOVpNKY6AqeKWWcgJwNejrQREfNY
/xHzGR2zvhFYcO7x5V2lafPHaB9CT6lxaRBsNFYfQf7ajaNGeHKXhjHsnQnlW46a
ELWIGbLJ
=55wg
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `6d0233da06adba4dcb766f9a62ac736cf11c827239891f0b3af49ea1152fcac1  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e8929401086d0233da06adba4dcb766f9a62ac736cf11c827239891f0b3af49ea1152fcac1f010f0fda69d7ba0a0b4e2eb58e9a1a21f0408fff01044eb7829b1342a68a5a8ef361c825fc608f1045ed43688f00839dd2b6f65c2ce4d0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0104bca8fb0c87f5e1963452b262a763fd708f1045ed43688f008f5bb659d2c15fd610083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0105bdb4a960d998352589f7292f969e03408f1045ed43688f008094f91e8a5d473e50083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df0109589d836dc8a404491ab0260dc03b6ee08f1045ed43687f0084d695a07cc8c234f0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

File: src/test/fuzz/script_sign.cpp (hunk 3 above)
Comment (MarcoFalke) [reply src/test/fuzz/script_sign.cpp:134]: @practicalswift Are you going to create a pull with the branch or should I do it?

Comment (practicalswift) [reply src/test/fuzz/script_sign.cpp:134]: @MarcoFalke I plan to do a FuzzBench style test between the two variations (`T ConsumeDeserializable(…)` and `std::optional<T> ConsumeDeserializable(…)`) to see which one that helps us reach the highest coverage after 24 hours of fuzzing using `libFuzzer` (with each harness given `1/N` slice of the runtime) with an empty input corpus. Does that sound like a fair experimental setup that could guide us regarding the trade-offs we're facing here? If so, I'll do it :)

Comment (MarcoFalke) [reply src/test/fuzz/script_sign.cpp:134]: Sure, can't hurt to do a benchmark
